cdrtools:  update to version 3.02

RobK88 opened the pull request:
#### Description
cdrtools:  update to version 3.02

* Reformat Portfile
* Add "makefile 1.0" PortGroup
* Add "conflicts_build 1.0" PortGroup
* Add "codeberg 1.0" PortGroup
* Add codeberg.setup after name
* Update version to 3.02-${codeberg.version}
* Add revision number
* Add audio to categories
* Add myself as a maintainer
* Remove BSD license
* Remove homepage
* Remove master_sites
* Update the checksums
* Remove "use_bzip2"
* Delete the patch files - patch-include_schily_sha2.h and mkisofs.c.patch
* Add patch block to delete "-noclobber" flag in cdrecord/Makefile.dfl
* Add patch block to replace "/opt/schily" with "${prefix}" in the source code files
* Add "port_conflict_check" procedure
* Add "port_conflict_check" procedure call to check against build conflicts with cdrtools versions 0 to 3.01_1
* Remove "use_configure no"
* Add "universal_variant no" to Portfile
* Change compiler.blacklist to compiler.backlist-append
* Remove "configure.cflags-append"
* Add "-lintl" to configure.ldflags-append
* Change depends_build to depends_build-append
* Change depends_lib to depends_lib-append
* Remove build dependency - port:smake
* Add build dependency - port:smake
* Add build dependency - port:gettext
* Add build dependency - port:libiconv
* Add library dependency - port:gettext-runtime
* Add library dependency - port:libiconv
* Change build.cmd to smake
* Delete "build.jobs -1" from the Portfile
* Add "use_parallel_build      no"
* Add note regarding the disabling of parallel builds
* Remove CC, COPTX, LDOPTX and LDPATH from build.args
* Change build.args to build.args-append
* Add INS_BASE, INS_RBASE, CC_OPT to build.args-append in the Portfile
* Remove post-build {} code block
* Add INS_RBASE to destroot.destdir in the Portfile
* Reformat post-distroot code block
* Fix post-destroot block to delete unnecessary binaries, manpages and docs
* Delete livecheck related statements

**IMPORTANT - PLEASE READ:**  This PR for `cdrtools` will not build if the old version of `cdrtools`, version 3.01, currently in the MacPorts repo has been installed and is active. To address this issue, a procedure called `port_conflict_check` has been added which checks whether `cdrtools`, version 3.01_1 or earlier has been installed. (Thanks to @mascguy for the code).  Once `smake` is built, `cdrtools` can be activated again.

FYI -- The old version of `cdrtools`, version 3.01, in the MacPorts repo installs OLD libschily header files and libraries. It is the existence of these old header files and libraries which prevents this updated port for `cdrtools` from building successfully. So the previous version of `cdrtools` must be uninstalled or deactivated first.

This PR for `cdrtools` no longer installs the libshily header files and libraries since they are not needed to run the binaries.  

Please see the following related Pull Requests:

(smake)  https://github.com/macports/macports-ports/pull/18655
(star)      https://github.com/macports/macports-ports/pull/18616

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 10.13.6 17G14042 x86_64
Xcode 10.1 10B61

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

RobK88 commented:
Please note that all the static libraries and header files have been removed in this upgrade to the `cdrtools` port.  If you think they are needed, please let me know.

catap commented on sysutils/cdrtools/Portfile:
```diff
@@ -1,98 +1,211 @@
+
+depends_fetch           port:git
+
+fetch.type              git
```
You don't need it here. I suggest to use technic similar with https://github.com/macports/macports-ports/pull/18655

RobK88 replied:
@catap - Since I now know how I can test this port locally on my Mac running High Sierra to overcome the Lets Encrypt cert expired bug (see https://trac.macports.org/wiki/ProblemHotlist), I have just updated this port, `cdrtools` to download the source tarball instead of using `git` as the fetch method.

RobK88 commented:
@catap - As with the PR for `smake`, using `${git_tag}` in the macro definition for `version` does not work:

```
$ port outdated
The following installed ports are outdated:
cdrtools                       3.01_1 < 3.02-{git_tag}_0   
```
The solution is to eliminate `${git_tag}` in the `version` macro definition or just eliminate `${git_tag}` altogether and use `${livecheck.version}` instead.  I will try the latter.

RobK88 commented:
We also have another bigger problem.  One cannot upgrade an older version of `cdrtools` (such as v3.01) to the most recent version of `cdrtools` without first uninstalling or deactivating the previous version.

If one tries, the build will fail and one will see the same errors I was experiencing when I was trying to get the port for the most recent version of `smake` to work on my Mac running High Sierra.

e.g.

```
searchinpath.c:116:19: error: use of undeclared identifier 'SIP_ONLY_PATH'
        if ((file_mode & SIP_ONLY_PATH) == 0 &&
                         ^
searchinpath.c:125:33: error: use of undeclared identifier 'SIP_NO_STRIPBIN'
                                file_mode & (SIP_PLAIN_FILE|SIP_NO_STRIPBIN)
```

@Dave-Allured found the problem and suggested that I uninstall all my previsions versions of the Schily Tools programs like `cdrtools`, version 3.01.  Once I did that I was able to install the most recent recent of `smake` and the most recent recent of `cdrtools` using the updated port in this Pull Request.

The issue is with the previous version of `cdrtools` (which is current version in the MacPorts repo).  the previous version installs unnecessary header files for libschily which are just too old to be used when building the most recent version of `cdrtools`.

What do we do here.  Do we somehow display a notice to users when they try to upgrade `cdrtools` telling them to first deactivate or uninstall their previous version before upgrading or installing `cdrtools`?

catap commented:
> If there is a better way to deal with this issue, please let me know.

You may use `conflicts_build` PG for such case. For example mumps can't be build when another mumps is activated.

See: https://github.com/macports/macports-ports/blob/master/math/mumps/Portfile

RobK88 commented:
@capat - I keep learning something new.  I learned how to set a global constant in a Portfile with `set` and I learned three new commands `git.cmd`, `extract.rename` and now `conflicts_build`.  It has been a great leaning experience!

With `conflicts_build`, can one set a version?  (e.g. version 3.01 and older?)  I fixed the problem in this updated port.  `cdrtools`, version  3,02 no longer installs unneeded static libraries and header files.  So once you have successfully installed `cdrtools` version 3.02, one no longer needs to uninstall or deactivate it before installing a newer version.

RobK88 commented:
Nope.  Neither `conflicts_build   cdrtools@3.01` or `conflicts_build   cdrtools@3.01_1` work.  I guess I will just have to use `conflicts_build   cdrtools` in the Portfile.  And after a few newer releases, I may just remove `conflicts_build` altogether from the Portfile since most people would have already upgraded to a release greater than 3.01.

RobK88 commented:
P.S.  I suspect there is a way to only use `conflicts_build` with only certain releases by using a pre-build { } block and some tcl code.  Maybe something like:
```

pre-build {
      if { $version < 3.01 } {
             conflicts_build  cdrtools
     }
}
```

catap commented:
@RobK88 are you sure that such conflict won't appear in the future update?

RobK88 commented:
Yes.  For some reason, the old port for `cdrtools` installed a lot of bloat including the libschily header files and static libraries.  Both of which were needed to build the `cdrtools` like `cdrecord` and `cdda2wav`.  But were are not needed to run them afterwards.

Years ago, the developer created its own source package for `cdrtools` which has not been updated for years.  To get the most recent version of `cdrtools`, one must now use the source in the Schily Tools tarball.  In other words, one must now compile and install all the Schily Tools which also include the `cdrtools`.

So in this newly written port for `cdrtools`, the port builds all the binaries, libraries etc from the latest version of the Schily Tools source tarball and then deletes all the unneeded binaries, header files, libraries, manpages etc in `destroot`.  I deleted all the Schily library header files and libraries in `destroot`.  

It is the existence of the old header files and libraries in `/opt/local/include/schily` that causes the problem when building newer versions of `cdrtools`.  That should not be a problem when one upgrades `cdrtools` version 3.02 to a newer release since these header files and libraries should no longer exist.  But upgrading `cdrtools` v 3.01 or earlier to version 3.02 or newer is a different matter and will fail.

P.S.  FYI -- In this rewritten port, I deleted all the bloat in `destroot` that was created in `/opt/local/lib` and `/opt/local/include`.  None of the libraries created are dynamic so they are not used to actually run the `cdrtools` like `cdrecord`, `cdda2wav` etc.  If you think they really should be included, please let me know.  In my view, if developers want the latest Schily libraries and header files for a software project that they are developing, they can compile all the Schily Tools themselves (or maybe a new port should be created for all the Schily Tools.  That would be easy to do now).

catap commented:
@RobK88 I suggest to keep things simpler: just add `conflicts_build  cdrtools` with comment that it is transitional, and the next update should remove it, or in couple of months. Or a year if you would like to be on very safe side.

RobK88 commented:
Until I can figure out how to prevent `cdrtools` from building only when certain versions of the `cdrtools` port are present, I added the following to the Portfile for `cdrtools`:

PortGroup                           conflicts_build 1.0
conflicts_build-append              cdrtools

I also reformatted the post-destroot code block and replaced the Tcl commands with preferred MacPorts macros wherever possible in the the post-destroot code block. I do not think replacing file `delete -force` with just `delete` will make much difference. But apparently, it is much better to use the MacPorts macro `move` instead of the Tcl `file rename` since the MacPorts macro will correctly handle renames that only change the case of a file on a case-insensitive filesystem and the Tcl command may not.

RobK88 commented:
@CAPAT - Please see the updated description of the PR at the top of this webpage. I did my best to explain the situation regarding the problems created by the old incompatible versions of `smake` and `cdrtools` in the MacPorts repo.

catap commented:
@RobK88 after some thoughts I suggest a different approach.

Create a `Portfile` for a project `schilytools` with subports for each subproject from this repository. Like `smake` or `star`.

Broken or out of updated ports can be replaced by `schilytools-xxx` ports.

As nice side effect it allows to avoid a hack with `git_tag` and use simple version which is matched upstream git tag :)

What do you think?

catap commented:
Also, see https://github.com/macports/macports-ports/pull/18753 where I've introduced a PG for codeberg

RobK88 commented:
@catap - Many thanks for the great news. The "codeberg 1.0" PortGroup will defeiitely simplify things. I have rebased my `cdrtools` branch to the latest master. I have made the changes to the Portfile to use the codeberg PortGroup, rebased and force pushed.

RobK88 commented:
@catap - I am not sure how to turn my three PR's into one.  And to be honest I do not have the time for at least a week to turn my attention to such a task.  I will be travelling soon.

IMHO, the real issue is the PR for `smake`.   Once the `smake` port has been merged into the MacPorts repo, I can finalize the PR's for `star` and `cdrtools` by making `smake` a build dependency.  (With `smake` a build dependency, it will no longer be nececcsary for the `star` and `cdrtools` ports to build a boostrapped version of `smake`.  This will speed up the builds for `star` and `cdrtools` significantly).

RobK88 commented:
@catap - P.S.  I have also developed a port for all the Schily Tools.  (I did this to make sure I did not miss anything when developing the ports for `star` and `cdrtools`).  

The development of the Schily Tools port  was a LOT of work since the Schily Tools include a lot of tools with the SAME names as existing BSD and GNU tools (e.g.  cal, copy, compare, gnutar etc).  So I had to rename many of them with an 's' prefix (like was done with the GNU coreutils port).  And to be honest I do not know why Mac users would really want to make use of any of the Schily Tools other than `cdrtools` and `star`.  So I am not sure whether I will submit the port for all the Schily Tools.

catap commented:
> @catap - P.S. I have also developed a port for all the Schily Tools. (I did this to make sure I did not miss anything when developing the ports for `star` and `cdrtools`).
> 
> The development of the Schily Tools port was a LOT of work since the Schily Tools include a lot of tools with the SAME names as existing BSD and GNU tools (e.g. cal, copy, compare, gnutar etc). So I had to rename many of them with an 's' prefix (like was done with the GNU coreutils port). And to be honest I do not know why Mac users would really want to make use of any of the Schily Tools other than `cdrtools` and `star`. So I am not sure whether I will submit the port for all the Schily Tools.

Well.. This can be useful for some users. Specially if you make a lot of efforts to make such port I not sure that it is good idea to waste it.

For example I supports [beets](https://ports.macports.org/port/beets/details/) which isn't that popular and quite large if we speak about port size. And statistic says that at least 2 users are using it since I've introduced it :)

The same applies for [retdec](https://ports.macports.org/port/retdec/details/) which is also quite complicated and long to build, and which is very special software ;)

mascguy commented on sysutils/cdrtools/Portfile:
```diff
@@ -1,98 +1,205 @@
+                        integrity, and write them to a disc.
+
+homepage                https://codeberg.org/schilytools/schilytools
+master_sites            ${homepage}/archive/
```
pg `codeberg` sets `homepage` and `master_sites`. So go ahead and remove these two lines.

RobK88 replied:
Done!

mascguy commented on sysutils/cdrtools/Portfile:
```diff
@@ -1,98 +1,205 @@
+
+codeberg.setup          schilytools schilytools 2023-04-19
+
+name                    cdrtools
```
Ensure `name` is set prior to `codeberg.setup`, to ensure `dist_subdir` and friends are set correctly.

RobK88 replied:
Done.

mascguy commented on sysutils/cdrtools/Portfile:
```diff
@@ -1,98 +1,205 @@
+                          port:libiconv
+
+depends_build-append    port:gmake \
+                        port:gettext-runtime \
```
Change the build dep to `gettext`.

RobK88 replied:
Done.

mascguy commented on sysutils/cdrtools/Portfile:
```diff
@@ -1,98 +1,205 @@
+
+depends_build-append    port:gmake \
+                        port:gettext-runtime \
+                        port:libiconv
```
Since `libiconv` is already a lib dep, you don't need to declare a build dep for it.

mascguy replied:
> Since `libiconv` is already a lib dep, you don't need to declare a build dep for it.

it looks like this was the only thing you missed, with your latest push.

RobK88 replied:
Are you sure?  I see "localization" of various files during both the build and destroot stages.  

I soon discovered that I needed to add the "-lintl" flag to "configure.ldflags-append".  Otherwise the build would fail on Big Sur and above.

Are you telling me that I just need the "-intl" flag?  And I do not need the "libiconv" as a build dependency?  Is that right?

mascguy replied:
> Are you sure? I see "localization" of various files during both the build and destroot stages.
> 
> I soon discovered that I needed to add the "-lintl" flag to "configure.ldflags-append". Otherwise the build would fail on Big Sur and above.
> 
> Are you telling me that I just need the "-intl" flag? And I do not need the "libiconv" as a build dependency? Is that right?

In short, all lib deps are also build deps. So if it's declared as a lib dep, it will also be pulled in for the build. (You'll still need your linker flag though, obviously.)

mascguy commented on sysutils/cdrtools/Portfile:
```diff
@@ -1,98 +1,205 @@
+# the rest of the port.
+#
+
+build.cmd               ${worksrcpath}/Gmake
```
Just to verify, is this the correct case for the bundled `gmake`? (If not, this will fail on our buildbots, as we use a case-sensitive filesystem there.)

mascguy replied:
> Just to verify, is this the correct case for the bundled `gmake`? (If not, this will fail on our buildbots, as we use a case-sensitive filesystem there.)

Checked the distfiles, and sure enough, `Gmake` is correct. All good!

RobK88 replied:
@macsguy -- I am updating this port given all your great feedback on the `smake` PR.  And yes the capital G in `./Gmake` is correct.  It is a script provided by the developer.  In any event, this statement will change!  since you just merged the `smake` PR.  

Because MacPorts now has an up to date version of `smake` in its repo,  this line will change to `build.cmd   smake`.   The `port.smake` is also being added as a build dependency and `port:gmake` is being removed.  I will do a rebase and force push VERY soon.

RobK88 commented:
@mascguy - Just did a force push incorporating all your suggested changed from the previous `smake` port.  This PR for `cdrtools` has also been updated to use the recently merged `smake` port to build all the `cdrtools`.  The Portfile should be easier to understand now.  I am currently building the +univesral variant to see if that will work.

RobK88 commented:
The universal build failed.  I guess I will need to remove the "muniversal 1.1" PortGroup and add "universal_variant no".

catap commented:
> The universal build failed. I guess I will need to remove the `muniversal 1.1` PortGroup and add `universal_variant no` to the Portfile

I may take a look on this and suggest a patch; but in an hour or two

RobK88 commented:
I cannot even build the non-universal x86_64 binaries now.  The build fails in the install stage.  Before I could build the +universal variant of `cdrtools`, I had to install the `+universal` variant of `smake`.  Maybe it is the `+universal` variant of `smake` that is the problem....  Will see.  Time for testing.

mascguy reviewed (approved):
Looks great at this point. But we'll wait to merge, until you're done with your Universal testing.

RobK88 commented:
@mascguy - Thanks!  I hope to be finished testing later tonight.  Making +universal builds is VERY slow.  I will let you know when I am done.

RobK88 commented:
@catap - No.  I can build the +universal variant for `smake` which is needed to build the +universal variant for `cdrtools`.  But I cannot build the +universal variant for `cdrtools` using the `muniversal 1.1` Portgroup.  It fails every time.  See attached log.

[cdrtools universal variant Build Log.txt](https://github.com/macports/macports-ports/files/11559581/cdrtools.universal.variant.Build.Log.txt)

catap commented:
@RobK88 the log file contins:
```
--->  Running post destroot cdrtools for architecture x86_64
Error: Failed to destroot cdrtools: no files matched glob pattern "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_sysutils_cdrtools/cdrtools/work/destroot-x86_64/opt/local/bin/*"
```
that means that your `post-destroot` phase failed. My bet that it fails on line:
```
                        set allPrograms [ glob ${destroot}${prefix}/bin/* ]
```

RobK88 commented:
> that means that your `post-destroot` phase failed. My bet that it fails on line:
> 
> ```
>                         set allPrograms [ glob ${destroot}${prefix}/bin/* ]
> ```

Yep.  You are most likely right.  We might be able to fix it but we would probably have to add a lot of code in `post-destroot` when the universal variant is being built. (i.e.  `if [variant_isset "universal"] {  lots of new code.... }`

Or maybe you can suggest an easier fix.  Or we just go with `universal_variant no` in the Portfile.  

Let me build the `+universal` variant again for `cdrtools` and see exactly what is being built in `destroot`.

catap commented:
@RobK88 or you may simplify that by installing everything into `${prefix}/libexec/cdrtools` and adding a symlinks from `${prefix}/bin/cdda2mp3` to `${prefix}/libexec/cdrtools/bin/cdda2mp3` and so on.

RobK88 commented:
@catap -- It is not the post-destroot phase that failed.  It is the destroot phase that is supposed to create the universal binaries that is failing.  The `munivesral` PortGroup creates a numbers of folders in the ${worksrcdir}:

```
destroot
destroot-i386
destroot-x86-64
```

If you look in the `/opt/local/bin` sub-directories inside the `destroot-i386` or `destroot-x86-64` folders, you will see NO binaries.  But if you look in the `/opt/local/bin` folder inside the `destroot` folder, you will see binaries.  But all of them are just 64 bit.  No way to create universal binaries. 

IMHO, the `muniversal` PortGroup is juts not working right with this port.  

You also see a number of warnings that you ONLY see when making the universal build:
`NOTICE: Partial source (SRCROOT/lib) missing`

My recommendation:  Just add `universal_variant no` to the Portfile for `cdrtools`.

RobK88 commented:
P.S.  If you inside the `${destroot}/2023-04-19-i386` folder, you also only see 64 bit binaries -- NO 32-bit binaries.  It does not look like any 32 binaries are being built.   It is probably a problem with `smake` not providing the right `${arch}` flag to the compiler.  I might be able to fix it by providing the arch info to the compiler via a `$COPTX` in `build.args-append`.

I would leave `universal_variant no` in the Portfile for now.  I will see if I can fix it next week when I am back home.  I have to start packing for my trip back North.

mascguy commented:
> I would leave `universal_variant no` in the Portfile for now. I will see if I can fix it next week when I am back home. I have to start packing for my trip back North.

Works for me! I'll go ahead and merge, and you can revisit on your return.

Great work on this Rob!

RobK88 commented:
@mascgiy -- Thanks.  

P.S.  I doubt adding the following to `build.args-append` would help with the universal build.:

```
COPTX="${configure.cflags} [get_canonical_archflags cc]" \
LDOPTX="${configure.ldflags} [get_canonical_archflags ld]"
```

Why?  The archflag is NOT set when you try to install a port with the +universal variant.  You will see:

```
sudo port -v install cdrtools +universal
Checking for conflict against port: cdrtools
Error: universal: merge: architecture is not set
Error: Unable to open port: unknown architecture (empty)
```
You probably could fix it by hacking the `muniversal` PortGroup to ensure that the right arch flag is being sent to `smake` during universal builds.  Maybe create a new PortGroup called `smakeFixUniversal` that will do the trick.

catap commented:
Ah, its clear.

Let see if someone really need that tools as universal and I agree that switching it requires too much efforts and probably for nothing.

RobK88 commented:
I have been able to make this port universal.  When I get a chance I will submit a PR next week.

RobK88 commented:
FYI -- I submitted a PR for the +universal variant for `cdrtools`
see
https://github.com/macports/macports-ports/pull/20704
